cpp/ql/src/Critical/MissingCheckScanf.ql

```diff
@@ -20,81 +20,187 @@ import semmle.code.cpp.dataflow.DataFlow
+    isSink(instr, _, _)
+    or
+    exists(Instruction succ | revFlow0(succ) | instr.getASuccessor() = succ)
+  )
+}
+
+/**
+ * Holds if `instr` is part of a path from a call `to a `scanf`-like function
+ * that writes to a variable with value number `vn`, without passing through
+ * redefinitions of the variable.
+ */
+pragma[nomagic]
+private predicate fwdFlow(Instruction instr, ValueNumber vn) {
+  revFlow0(instr) and
+  (
+    isSource(_, _, instr, vn, _)
+    or
+    exists(Instruction prev |
+      fwdFlow(prev, vn) and
+      prev.getASuccessor() = instr and
+      not isBarrier(instr, vn)
+    )
+  )
+}
+
+/**
+ * Holds if `instr` is part of a path from a call `to a `scanf`-like function
+ * that writes to a variable with value number `vn`, without passing through
+ * redefinitions of the variable.
+ */
+pragma[nomagic]
+predicate revFlow(Instruction instr, ValueNumber vn) {
+  fwdFlow(instr, vn) and
+  (
+    isSink(instr, _, vn)
+    or
+    exists(Instruction succ | revFlow(succ, vn) |
+      instr.getASuccessor() = succ and
+      not isBarrier(succ, vn)
     )
-  }
+  )
+}
 
-  private Instruction bigStep() {
-    result = this.smallStep(instr)
+/**
+ * A type that bundles together a reachable instruction with the appropriate
+ * value number (i.e., the value number that's transferred from the source
+ * to the sink).
+ */
+newtype TNode = MkNode(Instruction instr, ValueNumber vn) { revFlow(instr, vn) }
+
+class Node extends MkNode {
+  ValueNumber vn;
+  Instruction instr;
+
+  Node() { this = MkNode(instr, vn) }
+
+  final string toString() { result = instr.toString() }
+
+  final Node getASuccessor() { result = MkNode(instr.getASuccessor(), vn) }
+
+  final Location getLocation() { result = instr.getLocation() }
+}
+
+/**
+ * Holds if `instr` is an instruction with value number `vn` that is
+ * used in a store operation, or is overwritten by another call to
+ * a `scanf`-like function.
+ */
+private predicate isBarrier(Instruction instr, ValueNumber vn) {
+  // We only need to compute barriers for instructions that we
+  // managed to hit during the initial flow stage.
+  revFlow0(pragma[only_bind_into](instr)) and
+  valueNumber(instr) = vn and
+  exists(Expr e | instr.getAst() = e |
+    instr = any(StoreInstruction s).getDestinationAddress()
     or
-    exists(Instruction i | i = this.bigStep() | result = this.smallStep(i))
-  }
-
-  private Instruction smallStep(Instruction i) {
-    instr.getASuccessor*() = i and
-    i.getASuccessor() = result and
-    not this.isBarrier(result)
-  }
-
-  private predicate isBarrier(Instruction i) {
-    valueNumber(i) = valNum and
-    exists(Expr e | i.getAst() = e |
-      i = any(StoreInstruction s).getDestinationAddress()
+    isSource(_, _, _, _, [e, e.getParent().(AddressOfExpr)])
+  )
+}
+
+/** Holds if `n1` steps to `n2` in a single step. */
+predicate isSuccessor(Node n1, Node n2) { n1.getASuccessor() = n2 }
+
+predicate hasFlow(Node n1, Node n2) = fastTC(isSuccessor/2)(n1, n2)
```

Comment:
Is there a difference between this and `isSuccessor*`?

Reply:
Yeah, this `fastTC` is equivalent to `isSuccessor+`.

`isSuccessor+` should _hopefully_ compile to exactly this, but there's no guarantee that it is 😄. It might also be compiled to a naive transitive closure predicate. I didn't actually check how it compiled the original code.